<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Scott Sherrill-Mix" />


<title>taxonomizr usage</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">taxonomizr usage</h1>
<h4 class="author">Scott Sherrill-Mix</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><code>taxonomizr</code> provides some simple functions to parse NCBI taxonomy files and accession dumps and efficiently use them to assign <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/">taxonomy</a> to accession numbers or taxonomic IDs. This is useful for example to assign taxonomy to BLAST results. This is all done locally after downloading the appropriate files from NCBI using included functions (see <a href="#preparation">below</a>).</p>
<p>The major functions are: * <code>prepareDatabase</code>: download data from NCBI and prepare SQLite database * <code>accessionToTaxa</code>: convert accession numbers to taxonomic IDs * <code>getTaxonomy</code>: convert taxonomic IDs to taxonomy</p>
<p>More specialized functions are: * <code>getId</code>: convert a biological name to taxonomic ID * <code>getAccessions</code>: find accessions for a given taxonomic ID * <code>makeNewick</code>: generate a Newick formatted tree from taxonomic output</p>
<p>And a simple use case might look like (see below for more details):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(taxonomizr)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="co">#note this will require a lot of hard drive space, bandwidth and time to process all the data from NCBI</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw">prepareDatabase</span>(<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a>
<a class="sourceLine" id="cb1-4" title="4">blastAccessions&lt;-<span class="kw">c</span>(<span class="st">&quot;Z17430.1&quot;</span>,<span class="st">&quot;Z17429.1&quot;</span>,<span class="st">&quot;X62402.1&quot;</span>) </a>
<a class="sourceLine" id="cb1-5" title="5">ids&lt;-<span class="kw">accessionToTaxa</span>(blastAccessions,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a>
<a class="sourceLine" id="cb1-6" title="6"><span class="kw">getTaxonomy</span>(ids,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a></code></pre></div>
</div>
<div id="requirements" class="section level2">
<h2>Requirements</h2>
<p>This package downloads a few databases from NCBI and stores them in an easily accessible form on the hard drive. This ends up taking a decent amount of space so you’ll probably want around 75 Gb of free hard drive space.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>The package is on CRAN, so it should install with a simple:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;taxonomizr&quot;</span>)</a></code></pre></div>
<p>If you want the development version directly from github, use the <a href="https://github.com/r-lib/devtools"><code>devtools</code></a> library and run:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;sherrillmix/taxonomizr&quot;</span>)</a></code></pre></div>
<p>To use the library, load it in R:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">library</span>(taxonomizr)</a></code></pre></div>
</div>
<div id="preparation" class="section level2">
<h2>Preparation<a name="preparation"></a></h2>
<p>Since version 0.5.0, there is a simple function to run all preparations. Note that you’ll need a bit of time, download bandwidth and hard drive space before running this command (we’re downloading taxonomic assignments for every record in NCBI). To create a SQLite database called <code>accessionTaxa.sql</code> in the current working directory (you may want to store this somewhere more centrally located so it does not need to be duplicated with every project), we can run:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw">prepareDatabase</span>(<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a></code></pre></div>
<pre><code>## Downloading names and nodes with getNamesAndNodes()</code></pre>
<pre><code>## Downloading accession2taxid with getAccession2taxid()</code></pre>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## Preprocessing names with read.names.sql()</code></pre>
<pre><code>## Preprocessing nodes with read.nodes.sql()</code></pre>
<pre><code>## Preprocessing accession2taxid with read.accession2taxid()</code></pre>
<pre><code>## Reading ./nucl_gb.accession2taxid.gz.</code></pre>
<pre><code>## Reading ./nucl_wgs.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<pre><code>## [1] &quot;accessionTaxa.sql&quot;</code></pre>
<p>If everything works then that should have prepared a SQLite database ready for use. You can skip the “Manual preparation” steps below.</p>
<p>All files are cached locally and so the preparation is only required once (delete/rename the SQLite database and recall the function to regenerate the database). It is not necessary to manually check for the presence of the database since the function checks to see if SQLite database is present and if so skips downloading/processing. For example, running the command again produces:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="kw">prepareDatabase</span>(<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a></code></pre></div>
<pre><code>## SQLite database accessionTaxa.sql already exists. Delete to regenerate</code></pre>
<pre><code>## [1] &quot;accessionTaxa.sql&quot;</code></pre>
</div>
<div id="assigning-taxonomy" class="section level2">
<h2>Assigning taxonomy</h2>
<div id="producing-accession-numbers" class="section level3">
<h3>Producing accession numbers</h3>
<p>NCBI accession numbers are often obtained when doing a BLAST search (usually the second column of output from blastn, blastx, blastp, …). For example the output might look like:</p>
<pre><code>read1   gi|326539903|gb|CP002582.1|     69.68   1745    448     69      3       1702    3517898 3519606 3e-169  608
read2   gi|160426828|gb|CP000885.1|     68.46   1763    452     82      3       1711    1790367 1788655 4e-140  511
...</code></pre>
<p>So to identify a taxon for a given sequence you would blast it against e.g. the NCBI nt database and load the results into R. For NCBI databases, the accession number is often the 4th item in the <code>|</code> (pipe) separated reference field (often the second column in a tab separated result). For example, the <code>CP002582.1</code> in the gi|326539903|gb|<strong>CP002582.1</strong>| above.</p>
<p>So just as an example, reading in blast results might look something like:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1">blastResults&lt;-<span class="kw">read.table</span>(<span class="st">&#39;XXXX.blast&#39;</span>,<span class="dt">header=</span><span class="ot">FALSE</span>,<span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb21-2" title="2"><span class="co">#grab the 4th |-separated field from the reference name in the second column</span></a>
<a class="sourceLine" id="cb21-3" title="3">accessions&lt;-<span class="kw">sapply</span>(<span class="kw">strsplit</span>(blastResults[,<span class="dv">2</span>],<span class="st">&#39;</span><span class="ch">\\</span><span class="st">|&#39;</span>),<span class="st">&#39;[&#39;</span>,<span class="dv">4</span>)</a></code></pre></div>
</div>
<div id="finding-taxonomy-for-ncbi-accession-numbers" class="section level3">
<h3>Finding taxonomy for NCBI accession numbers</h3>
<p>Now we are ready to convert NCBI accession numbers to taxonomic IDs. For example, to find the taxonomic IDs associated with NCBI accession numbers “LN847353.1” and “AL079352.3”:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1">taxaId&lt;-<span class="kw">accessionToTaxa</span>(<span class="kw">c</span>(<span class="st">&quot;LN847353.1&quot;</span>,<span class="st">&quot;AL079352.3&quot;</span>),<span class="st">&quot;accessionTaxa.sql&quot;</span>)</a>
<a class="sourceLine" id="cb22-2" title="2"><span class="kw">print</span>(taxaId)</a></code></pre></div>
<pre><code>## [1] 1313 9606</code></pre>
<p>And to get the taxonomy for those IDs:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1"><span class="kw">getTaxonomy</span>(taxaId,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a></code></pre></div>
<pre><code>##      superkingdom phylum       class      order            
## 1313 &quot;Bacteria&quot;   &quot;Firmicutes&quot; &quot;Bacilli&quot;  &quot;Lactobacillales&quot;
## 9606 &quot;Eukaryota&quot;  &quot;Chordata&quot;   &quot;Mammalia&quot; &quot;Primates&quot;       
##      family             genus           species                   
## 1313 &quot;Streptococcaceae&quot; &quot;Streptococcus&quot; &quot;Streptococcus pneumoniae&quot;
## 9606 &quot;Hominidae&quot;        &quot;Homo&quot;          &quot;Homo sapiens&quot;</code></pre>
<p>You can also get taxonomy for NCBI accession numbers without versions (the .X following the main number e.g. the “.1” in LN847353.1) using the <code>version=&#39;base&#39;</code> argument of <code>accessionToTaxa</code>:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1">taxaId&lt;-<span class="kw">accessionToTaxa</span>(<span class="kw">c</span>(<span class="st">&quot;LN847353&quot;</span>,<span class="st">&quot;AL079352&quot;</span>),<span class="st">&quot;accessionTaxa.sql&quot;</span>)</a>
<a class="sourceLine" id="cb26-2" title="2"><span class="kw">print</span>(taxaId)</a></code></pre></div>
<pre><code>## [1] NA NA</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1">taxaId&lt;-<span class="kw">accessionToTaxa</span>(<span class="kw">c</span>(<span class="st">&quot;LN847353&quot;</span>,<span class="st">&quot;AL079352&quot;</span>),<span class="st">&quot;accessionTaxa.sql&quot;</span>,<span class="dt">version=</span><span class="st">&#39;base&#39;</span>)</a>
<a class="sourceLine" id="cb28-2" title="2"><span class="kw">print</span>(taxaId)</a></code></pre></div>
<pre><code>## [1] 1313 9606</code></pre>
</div>
<div id="finding-taxonomy-for-taxonomic-names" class="section level3">
<h3>Finding taxonomy for taxonomic names</h3>
<p>If you’d like to find IDs for taxonomic names then you can do something like:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1">taxaId&lt;-<span class="kw">getId</span>(<span class="kw">c</span>(<span class="st">&#39;Homo sapiens&#39;</span>,<span class="st">&#39;Bos taurus&#39;</span>,<span class="st">&#39;Homo&#39;</span>,<span class="st">&#39;Alces alces&#39;</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a>
<a class="sourceLine" id="cb30-2" title="2"><span class="kw">print</span>(taxaId)</a></code></pre></div>
<pre><code>## [1] &quot;9606&quot; &quot;9913&quot; &quot;9605&quot; &quot;9852&quot;</code></pre>
<p>And again to get the taxonomy for those IDs use <code>getTaxonomy</code>:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1">taxa&lt;-<span class="kw">getTaxonomy</span>(taxaId,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a>
<a class="sourceLine" id="cb32-2" title="2"><span class="kw">print</span>(taxa)</a></code></pre></div>
<pre><code>
##      superkingdom phylum     class      order      family      genus  
## 9606 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot; 
## 9913 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA         &quot;Bovidae&quot;   &quot;Bos&quot;  
## 9605 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot; 
## 9852 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA         &quot;Cervidae&quot;  &quot;Alces&quot;
##      species       
## 9606 &quot;Homo sapiens&quot;
## 9913 &quot;Bos taurus&quot;  
## 9605 NA            
## 9852 &quot;Alces alces&quot;</code></pre>
</div>
<div id="condensing-taxonomy" class="section level3">
<h3>Condensing taxonomy</h3>
<p>You can use the <code>condenseTaxa</code> function to find the agreements among taxonomic hits. For example to condense the taxonomy from the previous section to the lowest taxonomic rank shared by all three taxa:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" title="1"><span class="kw">condenseTaxa</span>(taxa)</a></code></pre></div>
<pre><code>##   superkingdom phylum     class      order family genus species
## 1 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA    NA     NA    NA</code></pre>
<p>This function can also be fed a large number of grouped hits, e.g. BLAST hits for high throughput sequencing reads after filtering for the best hits for each read, and output a condensed taxonomy for each grouping:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" title="1">groupings&lt;-<span class="kw">c</span>(<span class="st">&#39;read1&#39;</span>,<span class="st">&#39;read2&#39;</span>,<span class="st">&#39;read1&#39;</span>,<span class="st">&#39;read2&#39;</span>)</a>
<a class="sourceLine" id="cb36-2" title="2"><span class="kw">condenseTaxa</span>(taxa,groupings)</a></code></pre></div>
<pre><code>##       superkingdom phylum     class      order      family      genus 
## read1 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot;
## read2 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA         NA          NA    
##       species
## read1 NA     
## read2 NA</code></pre>
</div>
<div id="finding-accessions-for-a-given-taxonomic-id" class="section level3">
<h3>Finding accessions for a given taxonomic ID</h3>
<p>To find all the accessions for a given taxonomic ID, you can use the <code>getAccessions</code> function. This is a bit of an unusual use case so to preserve space, an index is not created by default in <code>read.accession2taxid</code>. If you are going to use this function, you will want to rebuild the SQLite database with the <code>indexTaxa</code> argument set to true with something like:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1"><span class="kw">read.accession2taxid</span>(<span class="kw">list.files</span>(<span class="st">&#39;.&#39;</span>,<span class="st">&#39;accession2taxid.gz$&#39;</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>,<span class="dt">indexTaxa=</span><span class="ot">TRUE</span>,<span class="dt">overwrite=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## Reading nucl_gb.accession2taxid.gz.</code></pre>
<pre><code>## Reading nucl_wgs.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<p>Then you can get the accessions for taxa 3702 with a command like (note that the limit argument is used here in order to preserve space):</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1"><span class="kw">getAccessions</span>(<span class="dv">3702</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>,<span class="dt">limit=</span><span class="dv">10</span>)</a></code></pre></div>
<pre><code>##    taxa accession
## 1  3702  X58148.1
## 2  3702  X66414.1
## 3  3702  X60045.1
## 4  3702  X07376.1
## 5  3702  X54927.1
## 6  3702  X54926.1
## 7  3702  X54928.1
## 8  3702  X54930.1
## 9  3702  X54929.1
## 10 3702  X52320.1</code></pre>
</div>
<div id="convert-taxonomy-to-newick-tree" class="section level3">
<h3>Convert taxonomy to Newick tree</h3>
<p>This is probably only useful in a few specific cases but a convenience function <code>makeNewick</code> to convert taxonomy into a Newick tree is included. The function takes a matrix giving with columns corresponding to taxonomic categories and rows different to taxonomic assignments, e.g. the output from <code>condenseTaxa</code> or <code>getTaxonomy</code> and reduces it to a Newick formatted tree. For example:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1">taxa</a></code></pre></div>
<pre><code>##      [,1]        [,2]       [,3]       [,4]       [,5]        [,6]   
## [1,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot; 
## [2,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Pan&quot;  
## [3,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; NA         &quot;Cervidae&quot;  &quot;Alces&quot;</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" title="1"><span class="kw">makeNewick</span>(taxa)</a></code></pre></div>
<pre><code>## [1] &quot;((((((Homo,Pan)Hominidae)Primates,((Alces)Cervidae)_)Mammalia)Chordata)Eukaryota)&quot;</code></pre>
</div>
</div>
<div id="changelog" class="section level2">
<h2>Changelog</h2>
<div id="v0.5.3" class="section level3">
<h3>v0.5.3</h3>
<ul>
<li>Fix named vector bug in <code>accessionToTaxa</code></li>
<li>Add <code>makeNewick</code> function</li>
<li>Deal with default 60 second timeout for downloads in R</li>
</ul>
</div>
<div id="v0.5.3-1" class="section level3">
<h3>v0.5.3</h3>
<ul>
<li>Remove <code>nucl_est</code> and <code>nucl_gss</code> from defaults since NCBI folded them into <code>nucl_gb</code> and removed</li>
<li>Squash R:devel bug</li>
</ul>
</div>
<div id="v0.5.0" class="section level3">
<h3>v0.5.0</h3>
<ul>
<li>Transitioned from data.table to SQLite</li>
<li>Addeded convenience <code>prepareDatabase()</code> function</li>
<li>Squashed Windows testing errors</li>
</ul>
</div>
</div>
<div id="manual-preparation-of-database-usually-not-necessary" class="section level2">
<h2>Manual preparation of database (usually not necessary)</h2>
<p><strong>Note:</strong> Since version 0.5.0, it is usually not necessary to run the following manually, the function <code>prepareDatabase()</code> should do most of this automatically for you (see <a href="#preparation">above</a>).</p>
<p>In order to avoid constant internet access and slow APIs, the first step in using the package is to downloads all necessary files from NCBI. This uses a bit of disk space but makes future access reliable and fast.</p>
<p><strong>Note:</strong> It is not necessary to manually check for the presence of these files since the functions automatically check to see if their output is present and if so skip downloading/processing. Delete the local files if you would like to redownload or reprocess them.</p>
<div id="download-names-and-nodes" class="section level3">
<h3>Download names and nodes</h3>
<p>First, download the necessary names and nodes files from <a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/">NCBI</a>:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1"><span class="kw">getNamesAndNodes</span>()</a></code></pre></div>
<pre><code>## [1] &quot;./names.dmp&quot; &quot;./nodes.dmp&quot;</code></pre>
</div>
<div id="download-accession-to-taxa-files" class="section level3">
<h3>Download accession to taxa files</h3>
<p>Then download accession to taxa id conversion files from <a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/">NCBI</a>. <strong>Note:</strong> this is a pretty <em>big</em> download (several gigabytes):</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" title="1"><span class="co">#this is a big download</span></a>
<a class="sourceLine" id="cb51-2" title="2"><span class="kw">getAccession2taxid</span>()</a></code></pre></div>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## [1] &quot;./nucl_gb.accession2taxid.gz&quot;  &quot;./nucl_wgs.accession2taxid.gz&quot;</code></pre>
<p>If you would also like to identify protein accession numbers, also download the prot file from NCBI (again this is a <em>big</em> download):</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" title="1"><span class="co">#this is a big download</span></a>
<a class="sourceLine" id="cb54-2" title="2"><span class="kw">getAccession2taxid</span>(<span class="dt">types=</span><span class="st">&#39;prot&#39;</span>)</a></code></pre></div>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## [1] &quot;./prot.accession2taxid.gz&quot;</code></pre>
</div>
<div id="convert-names-nodes-and-accessions-to-database" class="section level3">
<h3>Convert names, nodes and accessions to database</h3>
<p>Then process the downloaded names and nodes files into a more easily accessed form:</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb57-1" title="1"><span class="kw">read.names.sql</span>(<span class="st">&#39;names.dmp&#39;</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a>
<a class="sourceLine" id="cb57-2" title="2"><span class="kw">read.nodes.sql</span>(<span class="st">&#39;nodes.dmp&#39;</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a></code></pre></div>
<p>Next process the downloaded accession files into the same database (this one could take a while):</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" title="1"><span class="kw">read.accession2taxid</span>(<span class="kw">list.files</span>(<span class="st">&#39;.&#39;</span>,<span class="st">&#39;accession2taxid.gz$&#39;</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</a></code></pre></div>
<pre><code>## Reading nucl_gb.accession2taxid.gz.</code></pre>
<pre><code>## Reading nucl_wgs.accession2taxid.gz.</code></pre>
<pre><code>## Reading prot.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<p>Now everything should be ready for processing. All files are cached locally and so the preparation is only required once (or whenever you would like to update the data). It is not necessary to manually check for the presence of these files since the functions automatically check to see if their output is present and if so skip downloading/processing. Delete the local files if you would like to redownload or reprocess them.</p>
</div>
</div>
<div id="switch-from-data.table-to-sqlite" class="section level2">
<h2>Switch from data.table to SQLite</h2>
<p>Version 0.5.0 marked a change for name and node lookups from using data.table to using SQLite. This was necessary to increase performance (10-100x speedup for <code>getTaxonomy</code>) and create a simpler interface (a single SQLite database contains all necessary data). Unfortunately, this switch requires a couple breaking changes: * <code>getTaxonomy</code> changes from <code>getTaxonomy(ids,namesDT,nodesDT)</code> to <code>getTaxonomy(ids,sqlFile)</code> * <code>getId</code> changes from <code>getId(taxa,namesDT)</code> to <code>getId(taxa,sqlFile)</code> * <code>read.names</code> is deprecated, instead use <code>read.names.sql</code>. For example, instead of calling <code>names&lt;-read.names(&#39;names.dmp&#39;)</code> in every session, simply call <code>read.names.sql(&#39;names.dmp&#39;,&#39;accessionTaxa.sql&#39;)</code> once (or use the convenient <code>prepareDatabase</code> as <a href="#preparation">above</a>)). * <code>read.nodes</code> is deprecated, instead use <code>read.names.sql</code>. For example. instead of calling <code>nodes&lt;-read.names(&#39;nodes.dmp&#39;)</code> in every session, simply call <code>read.nodes.sql(&#39;nodes.dmp&#39;,&#39;accessionTaxa.sql&#39;)</code> once (or use the convenient <code>prepareDatabase</code> as <a href="#preparation">above</a>).</p>
<p>I’ve tried to ease any problems with this by overloading <code>getTaxonomy</code> and <code>getId</code> to still function (with a warning) if passed a data.table names and nodes argument and providing a simpler <code>prepareDatabase</code> function for completing all setup steps (hopefully avoiding direct calls to <code>read.names</code> and <code>read.nodes</code> for most users).</p>
<p>I plan to eventually remove data.table functionality to avoid a split codebase so please switch to the new SQLite format in all new code.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
