<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Scott Sherrill-Mix" />


<title>taxonomizr usage</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">taxonomizr usage</h1>
<h4 class="author">Scott Sherrill-Mix</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><code>taxonomizr</code> provides some simple functions to parse NCBI
taxonomy files and accession dumps and efficiently use them to assign <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/">taxonomy</a>
to accession numbers or taxonomic IDs. This is useful for example to
assign taxonomy to BLAST results. This is all done locally after
downloading the appropriate files from NCBI using included functions
(see <a href="#preparation">below</a>).</p>
<p>The major functions are: * <code>prepareDatabase</code>: download
data from NCBI and prepare SQLite database *
<code>accessionToTaxa</code>: convert accession numbers to taxonomic IDs
* <code>getTaxonomy</code>: convert taxonomic IDs to taxonomy</p>
<p>More specialized functions are: * <code>getId</code>: convert a
biological name to taxonomic ID * <code>getRawTaxonomy</code>: find all
taxonomic ranks for a taxonomic ID * <code>normalizeTaxa</code>: combine
raw taxonomies with different taxonomic ranks *
<code>condenseTaxa</code>: condense a set of taxa to their most specific
common branch * <code>makeNewick</code>: generate a Newick formatted
tree from taxonomic output * <code>getAccessions</code>: find accessions
for a given taxonomic ID * <code>getDescendants</code>: find descendants
for a given taxonomic ID</p>
<p>And a simple use case might look like (see below for more
details):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(taxonomizr)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#note this will require a lot of hard drive space, bandwidth and time to process all the data from NCBI</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">prepareDatabase</span>(<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>blastAccessions<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;Z17430.1&quot;</span>,<span class="st">&quot;Z17429.1&quot;</span>,<span class="st">&quot;X62402.1&quot;</span>) </span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>ids<span class="ot">&lt;-</span><span class="fu">accessionToTaxa</span>(blastAccessions,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">getTaxonomy</span>(ids,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
</div>
<div id="requirements" class="section level2">
<h2>Requirements</h2>
<p>This package downloads a few databases from NCBI and stores them in
an easily accessible form on the hard drive. This ends up taking a
decent amount of space so you’ll probably want around 75 Gb of free hard
drive space.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>The package is on CRAN, so it should install with a simple:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;taxonomizr&quot;</span>)</span></code></pre></div>
<p>If you want the development version directly from github, use the <a href="https://github.com/r-lib/devtools"><code>devtools</code></a>
library and run:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;sherrillmix/taxonomizr&quot;</span>)</span></code></pre></div>
<p>To use the library, load it in R:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(taxonomizr)</span></code></pre></div>
</div>
<div id="preparation" class="section level2">
<h2>Preparation<a name="preparation"></a></h2>
<p>Since version 0.5.0, there is a simple function to run all
preparations. Note that you’ll need a bit of time, download bandwidth
and hard drive space before running this command (we’re downloading
taxonomic assignments for every record in NCBI). To create a SQLite
database called <code>accessionTaxa.sql</code> in the current working
directory (you may want to store this somewhere more centrally located
so it does not need to be duplicated with every project), we can
run:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">prepareDatabase</span>(<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>## Downloading names and nodes with getNamesAndNodes()</code></pre>
<pre><code>## Downloading accession2taxid with getAccession2taxid()</code></pre>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## Preprocessing names with read.names.sql()</code></pre>
<pre><code>## Preprocessing nodes with read.nodes.sql()</code></pre>
<pre><code>## Preprocessing accession2taxid with read.accession2taxid()</code></pre>
<pre><code>## Reading ./nucl_gb.accession2taxid.gz.</code></pre>
<pre><code>## Reading ./nucl_wgs.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<pre><code>## [1] &quot;accessionTaxa.sql&quot;</code></pre>
<p>If everything works then that should have prepared a SQLite database
ready for use. You can skip the “Manual preparation” steps below.</p>
<p>All files are cached locally and so the preparation is only required
once (delete/rename the SQLite database and recall the function to
regenerate the database). It is not necessary to manually check for the
presence of the database since the function checks to see if SQLite
database is present and if so skips downloading/processing. For example,
running the command again produces:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">prepareDatabase</span>(<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>## SQLite database accessionTaxa.sql already exists. Delete to regenerate</code></pre>
<pre><code>## [1] &quot;accessionTaxa.sql&quot;</code></pre>
<p>Note that if you only want the taxonomic data and do not want to
assign taxonomy to accession ID then you can just get the much smaller
<code>names</code> and <code>nodes</code> data sets and exclude the
large download and time consuming databasing of accession IDs by setting
<code>getAccessions=FALSE</code> e.g.:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">prepareDatabase</span>(<span class="at">getAccessions=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## Downloading names and nodes with getNamesAndNodes()
##  [100%] Downloaded 57373562 bytes...
##  [100%] Downloaded 49 bytes...
## Preprocessing names with read.names.sql()
## Preprocessing nodes with read.nodes.sql()
## [1] &quot;nameNode.sqlite&quot;</code></pre>
<p>And if you area assigning taxonomy to protein data, then you would
want to grab the <code>prot.accession2taxid.gz</code> from NCBI by
specifying the <code>types=&#39;prot&#39;</code> argument (or
<code>types=c(&quot;nucl_gb&quot;, &quot;nucl_wgs&quot;,&quot;prot&quot;)</code> for proteins and
nucleotides):</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">prepareDatabase</span>(<span class="at">types=</span><span class="st">&#39;prot&#39;</span>)</span></code></pre></div>
<pre><code>## Downloading names and nodes with getNamesAndNodes()</code></pre>
<pre><code>## Preprocessing names with read.names.sql()</code></pre>
<pre><code>## Preprocessing nodes with read.nodes.sql()</code></pre>
<pre><code>## Downloading accession2taxid with getAccession2taxid()</code></pre>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## Preprocessing accession2taxid with read.accession2taxid()</code></pre>
<pre><code>## Reading ./prot.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<pre><code>## [1] &quot;nameNode.sqlite&quot;</code></pre>
</div>
<div id="assigning-taxonomy" class="section level2">
<h2>Assigning taxonomy</h2>
<div id="producing-accession-numbers" class="section level3">
<h3>Producing accession numbers</h3>
<p>NCBI accession numbers are often obtained when doing a BLAST search
(usually the second column of output from blastn, blastx, blastp, …).
For example the output might look like:</p>
<pre><code>read1   gi|326539903|gb|CP002582.1|     69.68   1745    448     69      3       1702    3517898 3519606 3e-169  608
read2   gi|160426828|gb|CP000885.1|     68.46   1763    452     82      3       1711    1790367 1788655 4e-140  511
...</code></pre>
<p>So to identify a taxon for a given sequence you would blast it
against e.g. the NCBI nt database and load the results into R. For NCBI
databases, the accession number is often the 4th item in the
<code>|</code> (pipe) separated reference field (often the second column
in a tab separated result). For example, the <code>CP002582.1</code> in
the gi|326539903|gb|<strong>CP002582.1</strong>| above.</p>
<p>So just as an example, reading in blast results might look something
like:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a>blastResults<span class="ot">&lt;-</span><span class="fu">read.table</span>(<span class="st">&#39;XXXX.blast&#39;</span>,<span class="at">header=</span><span class="cn">FALSE</span>,<span class="at">stringsAsFactors=</span><span class="cn">FALSE</span>)</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="co">#grab the 4th |-separated field from the reference name in the second column</span></span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>accessions<span class="ot">&lt;-</span><span class="fu">sapply</span>(<span class="fu">strsplit</span>(blastResults[,<span class="dv">2</span>],<span class="st">&#39;</span><span class="sc">\\</span><span class="st">|&#39;</span>),<span class="st">&#39;[&#39;</span>,<span class="dv">4</span>)</span></code></pre></div>
</div>
<div id="finding-taxonomy-for-ncbi-accession-numbers" class="section level3">
<h3>Finding taxonomy for NCBI accession numbers</h3>
<p>Now we are ready to convert NCBI accession numbers to taxonomic IDs.
For example, to find the taxonomic IDs associated with NCBI accession
numbers “LN847353.1” and “AL079352.3”:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a>taxaId<span class="ot">&lt;-</span><span class="fu">accessionToTaxa</span>(<span class="fu">c</span>(<span class="st">&quot;LN847353.1&quot;</span>,<span class="st">&quot;AL079352.3&quot;</span>),<span class="st">&quot;accessionTaxa.sql&quot;</span>)</span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a><span class="fu">print</span>(taxaId)</span></code></pre></div>
<pre><code>## [1] 1313 9606</code></pre>
<p>And to get the taxonomy for those IDs:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">getTaxonomy</span>(taxaId,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>##      superkingdom phylum       class      order            
## 1313 &quot;Bacteria&quot;   &quot;Firmicutes&quot; &quot;Bacilli&quot;  &quot;Lactobacillales&quot;
## 9606 &quot;Eukaryota&quot;  &quot;Chordata&quot;   &quot;Mammalia&quot; &quot;Primates&quot;       
##      family             genus           species                   
## 1313 &quot;Streptococcaceae&quot; &quot;Streptococcus&quot; &quot;Streptococcus pneumoniae&quot;
## 9606 &quot;Hominidae&quot;        &quot;Homo&quot;          &quot;Homo sapiens&quot;</code></pre>
<p>You can also get taxonomy for NCBI accession numbers without versions
(the .X following the main number e.g. the “.1” in LN847353.1) using the
<code>version=&#39;base&#39;</code> argument of
<code>accessionToTaxa</code>:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a>taxaId<span class="ot">&lt;-</span><span class="fu">accessionToTaxa</span>(<span class="fu">c</span>(<span class="st">&quot;LN847353&quot;</span>,<span class="st">&quot;AL079352&quot;</span>),<span class="st">&quot;accessionTaxa.sql&quot;</span>)</span>
<span id="cb39-2"><a href="#cb39-2" tabindex="-1"></a><span class="fu">print</span>(taxaId)</span></code></pre></div>
<pre><code>## [1] NA NA</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a>taxaId<span class="ot">&lt;-</span><span class="fu">accessionToTaxa</span>(<span class="fu">c</span>(<span class="st">&quot;LN847353&quot;</span>,<span class="st">&quot;AL079352&quot;</span>),<span class="st">&quot;accessionTaxa.sql&quot;</span>,<span class="at">version=</span><span class="st">&#39;base&#39;</span>)</span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a><span class="fu">print</span>(taxaId)</span></code></pre></div>
<pre><code>## [1] 1313 9606</code></pre>
</div>
<div id="finding-taxonomy-for-taxonomic-names" class="section level3">
<h3>Finding taxonomy for taxonomic names</h3>
<p>If you’d like to find IDs for taxonomic names then you can do
something like:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a>taxaId<span class="ot">&lt;-</span><span class="fu">getId</span>(<span class="fu">c</span>(<span class="st">&#39;Homo sapiens&#39;</span>,<span class="st">&#39;Bos taurus&#39;</span>,<span class="st">&#39;Homo&#39;</span>,<span class="st">&#39;Alces alces&#39;</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span>
<span id="cb43-2"><a href="#cb43-2" tabindex="-1"></a><span class="fu">print</span>(taxaId)</span></code></pre></div>
<pre><code>## [1] &quot;9606&quot; &quot;9913&quot; &quot;9605&quot; &quot;9852&quot;</code></pre>
<p>And again to get the taxonomy for those IDs use
<code>getTaxonomy</code>:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a>taxa<span class="ot">&lt;-</span><span class="fu">getTaxonomy</span>(taxaId,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span>
<span id="cb45-2"><a href="#cb45-2" tabindex="-1"></a><span class="fu">print</span>(taxa)</span></code></pre></div>
<pre><code>
##      superkingdom phylum     class      order      family      genus  
## 9606 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot; 
## 9913 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA         &quot;Bovidae&quot;   &quot;Bos&quot;  
## 9605 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot; 
## 9852 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA         &quot;Cervidae&quot;  &quot;Alces&quot;
##      species       
## 9606 &quot;Homo sapiens&quot;
## 9913 &quot;Bos taurus&quot;  
## 9605 NA            
## 9852 &quot;Alces alces&quot;</code></pre>
</div>
<div id="finding-descendants-for-a-given-taxa" class="section level3">
<h3>Finding descendants for a given taxa</h3>
<p>The function <code>getDescents</code> can be used to find all the
descendants at a taxonomic level for a given taxa. For example to find
all species (the default) in the Homininae subfamily (taxonomic ID
207598):</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="fu">getDescendants</span>(<span class="dv">207598</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Gorilla gorilla&quot;                   &quot;Gorilla beringei&quot;                 
## [3] &quot;Pan paniscus&quot;                      &quot;Pan troglodytes&quot;                  
## [5] &quot;Homo sapiens&quot;                      &quot;Homo heidelbergensis&quot;             
## [7] &quot;Homo sapiens environmental sample&quot; &quot;Homo sp.&quot;</code></pre>
<p>Or all genuses:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">getDescendants</span>(<span class="dv">207598</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>,<span class="st">&#39;genus&#39;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Gorilla&quot; &quot;Pan&quot;     &quot;Homo&quot;</code></pre>
<p>Note that an index for the nodes table was added in v0.10.1 to make
this run faster. If your database was created prior to v0.10.1 and you
need maximum speed for finding descendants then then please regenerate
the database.</p>
</div>
<div id="finding-common-names-for-taxonomic-ids" class="section level3">
<h3>Finding common names for taxonomic IDs</h3>
<p>If you’d like to find all common and other types of names for a given
taxa ID then you can use <code>getCommon</code>:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">getCommon</span>(<span class="fu">c</span>(<span class="dv">9913</span>,<span class="dv">9606</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>## [[1]]
##                         name                type
## 1                  Bos bovis             synonym
## 2     Bos primigenius taurus             synonym
## 3  Bos taurus Linnaeus, 1758           authority
## 4                 Bos taurus     scientific name
## 5      Bovidae sp. Adi Nefas            includes
## 6                     bovine         common name
## 7                     cattle genbank common name
## 8                        cow         common name
## 9                  dairy cow         common name
## 10           domestic cattle         common name
## 11              domestic cow         common name
## 12                        ox         common name
## 13                      oxen         common name
## 
## [[2]]
##                          name                type
## 1 Homo sapiens Linnaeus, 1758           authority
## 2                Homo sapiens     scientific name
## 3                       human genbank common name</code></pre>
<p>Or specify only a certain type(s) of name (“common” names seem to
often be split between “common name” and “genbank common name”):</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">getCommon</span>(<span class="fu">c</span>(<span class="dv">9913</span>,<span class="dv">9606</span>,<span class="dv">9894</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>,<span class="fu">c</span>(<span class="st">&#39;genbank common name&#39;</span>,<span class="st">&#39;common name&#39;</span>))</span></code></pre></div>
<pre><code>## [[1]]
##              name                type
## 1          bovine         common name
## 2          cattle genbank common name
## 3             cow         common name
## 4       dairy cow         common name
## 5 domestic cattle         common name
## 6    domestic cow         common name
## 7              ox         common name
## 8            oxen         common name
## 
## [[2]]
##    name                type
## 1 human genbank common name
## 
## [[3]]
##      name                type
## 1 giraffe genbank common name</code></pre>
<p>Note that databases created with <code>taxonomizr</code> versions
earlier than v0.9.4 do not contain the <code>type</code> field and so
the database will have to be reloaded to use this function. For example,
this could be done by calling:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a>taxonomizr<span class="sc">::</span><span class="fu">getNamesAndNodes</span>()</span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a>taxonomizr<span class="sc">::</span><span class="fu">read.names.sql</span>(<span class="st">&#39;names.dmp&#39;</span>,<span class="st">&#39;nameNode.sqlite&#39;</span>,<span class="at">overwrite=</span><span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="condensing-taxonomy-a.k.a.-lowest-common-ancestor-lca" class="section level3">
<h3>Condensing taxonomy a.k.a. lowest common ancestor LCA</h3>
<p>You can use the <code>condenseTaxa</code> function to find the
agreements among taxonomic hits. For example to condense the taxonomy
from the previous section to the lowest taxonomic rank shared by all
three taxa:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">condenseTaxa</span>(taxa)</span></code></pre></div>
<pre><code>##   superkingdom phylum     class      order family genus species
## 1 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA    NA     NA    NA</code></pre>
<p>This function can also be fed a large number of grouped hits,
e.g. BLAST hits for high throughput sequencing reads after filtering for
the best hits for each read, and output a condensed taxonomy for each
grouping:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a>groupings<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&#39;read1&#39;</span>,<span class="st">&#39;read2&#39;</span>,<span class="st">&#39;read1&#39;</span>,<span class="st">&#39;read2&#39;</span>)</span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a><span class="fu">condenseTaxa</span>(taxa,groupings)</span></code></pre></div>
<pre><code>##       superkingdom phylum     class      order      family      genus 
## read1 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot;
## read2 &quot;Eukaryota&quot;  &quot;Chordata&quot; &quot;Mammalia&quot; NA         NA          NA    
##       species
## read1 NA     
## read2 NA</code></pre>
</div>
<div id="find-all-taxonomic-assignments-for-a-given-taxa" class="section level3">
<h3>Find all taxonomic assignments for a given taxa</h3>
<p>To get all taxonomic assignments for a given taxa regardless of their
particular rank, you can use the <code>getRawTaxonomy</code> function.
Note that there are often many intermediate ranks outside the more
common taxonomic ranks. The function returns a list since different IDs
can have differing numbers of ranks. It is used similarly to
<code>getTaxonomy</code>:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="fu">getRawTaxonomy</span>(<span class="fu">c</span>(<span class="dv">9606</span>,<span class="dv">9913</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>## $`9606`
##                species                  genus              subfamily 
##         &quot;Homo sapiens&quot;                 &quot;Homo&quot;            &quot;Homininae&quot; 
##                 family            superfamily              parvorder 
##            &quot;Hominidae&quot;           &quot;Hominoidea&quot;           &quot;Catarrhini&quot; 
##             infraorder               suborder                  order 
##          &quot;Simiiformes&quot;          &quot;Haplorrhini&quot;             &quot;Primates&quot; 
##             superorder                  clade                clade.1 
##     &quot;Euarchontoglires&quot;        &quot;Boreoeutheria&quot;             &quot;Eutheria&quot; 
##                clade.2                  class                clade.3 
##               &quot;Theria&quot;             &quot;Mammalia&quot;              &quot;Amniota&quot; 
##                clade.4                class.1             superclass 
##            &quot;Tetrapoda&quot; &quot;Dipnotetrapodomorpha&quot;        &quot;Sarcopterygii&quot; 
##                clade.5                clade.6                clade.7 
##         &quot;Euteleostomi&quot;           &quot;Teleostomi&quot;        &quot;Gnathostomata&quot; 
##                clade.8              subphylum                 phylum 
##           &quot;Vertebrata&quot;             &quot;Craniata&quot;             &quot;Chordata&quot; 
##                clade.9               clade.10               clade.11 
##        &quot;Deuterostomia&quot;            &quot;Bilateria&quot;            &quot;Eumetazoa&quot; 
##                kingdom               clade.12           superkingdom 
##              &quot;Metazoa&quot;         &quot;Opisthokonta&quot;            &quot;Eukaryota&quot; 
##                no rank 
##   &quot;cellular organisms&quot; 
## 
## $`9913`
##                species                  genus              subfamily 
##           &quot;Bos taurus&quot;                  &quot;Bos&quot;              &quot;Bovinae&quot; 
##                 family             infraorder               suborder 
##              &quot;Bovidae&quot;               &quot;Pecora&quot;           &quot;Ruminantia&quot; 
##                  order             superorder                  clade 
##         &quot;Artiodactyla&quot;       &quot;Laurasiatheria&quot;        &quot;Boreoeutheria&quot; 
##                clade.1                clade.2                  class 
##             &quot;Eutheria&quot;               &quot;Theria&quot;             &quot;Mammalia&quot; 
##                clade.3                clade.4                class.1 
##              &quot;Amniota&quot;            &quot;Tetrapoda&quot; &quot;Dipnotetrapodomorpha&quot; 
##             superclass                clade.5                clade.6 
##        &quot;Sarcopterygii&quot;         &quot;Euteleostomi&quot;           &quot;Teleostomi&quot; 
##                clade.7                clade.8              subphylum 
##        &quot;Gnathostomata&quot;           &quot;Vertebrata&quot;             &quot;Craniata&quot; 
##                 phylum                clade.9               clade.10 
##             &quot;Chordata&quot;        &quot;Deuterostomia&quot;            &quot;Bilateria&quot; 
##               clade.11                kingdom               clade.12 
##            &quot;Eumetazoa&quot;              &quot;Metazoa&quot;         &quot;Opisthokonta&quot; 
##           superkingdom                no rank 
##            &quot;Eukaryota&quot;   &quot;cellular organisms&quot;</code></pre>
<p>These raw taxonomy with varying numbers of levels can be normalized
so that all taxa share the same number of levels (aligning by taxonomic
levels that are not the unspecific “clade”) using the
<code>normalizeTaxa</code> function:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a>raw<span class="ot">&lt;-</span><span class="fu">getRawTaxonomy</span>(<span class="fu">c</span>(<span class="dv">9606</span>,<span class="dv">9913</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a><span class="fu">normalizeTaxa</span>(raw)</span></code></pre></div>
<pre><code>##      no rank              superkingdom superkingdom.1 kingdom   kingdom.1  
## 9606 &quot;cellular organisms&quot; &quot;Eukaryota&quot;  &quot;Opisthokonta&quot; &quot;Metazoa&quot; &quot;Eumetazoa&quot;
## 9913 &quot;cellular organisms&quot; &quot;Eukaryota&quot;  &quot;Opisthokonta&quot; &quot;Metazoa&quot; &quot;Eumetazoa&quot;
##      kingdom.2   kingdom.3       phylum     subphylum  subphylum.1 
## 9606 &quot;Bilateria&quot; &quot;Deuterostomia&quot; &quot;Chordata&quot; &quot;Craniata&quot; &quot;Vertebrata&quot;
## 9913 &quot;Bilateria&quot; &quot;Deuterostomia&quot; &quot;Chordata&quot; &quot;Craniata&quot; &quot;Vertebrata&quot;
##      subphylum.2     subphylum.3  subphylum.4    superclass     
## 9606 &quot;Gnathostomata&quot; &quot;Teleostomi&quot; &quot;Euteleostomi&quot; &quot;Sarcopterygii&quot;
## 9913 &quot;Gnathostomata&quot; &quot;Teleostomi&quot; &quot;Euteleostomi&quot; &quot;Sarcopterygii&quot;
##      superclass.1           superclass.2 superclass.3 class      class.1 
## 9606 &quot;Dipnotetrapodomorpha&quot; &quot;Tetrapoda&quot;  &quot;Amniota&quot;    &quot;Mammalia&quot; &quot;Theria&quot;
## 9913 &quot;Dipnotetrapodomorpha&quot; &quot;Tetrapoda&quot;  &quot;Amniota&quot;    &quot;Mammalia&quot; &quot;Theria&quot;
##      class.2    class.3         superorder         order          suborder     
## 9606 &quot;Eutheria&quot; &quot;Boreoeutheria&quot; &quot;Euarchontoglires&quot; &quot;Primates&quot;     &quot;Haplorrhini&quot;
## 9913 &quot;Eutheria&quot; &quot;Boreoeutheria&quot; &quot;Laurasiatheria&quot;   &quot;Artiodactyla&quot; &quot;Ruminantia&quot; 
##      infraorder    parvorder    superfamily  family      subfamily   genus 
## 9606 &quot;Simiiformes&quot; &quot;Catarrhini&quot; &quot;Hominoidea&quot; &quot;Hominidae&quot; &quot;Homininae&quot; &quot;Homo&quot;
## 9913 &quot;Pecora&quot;      NA           NA           &quot;Bovidae&quot;   &quot;Bovinae&quot;   &quot;Bos&quot; 
##      species       
## 9606 &quot;Homo sapiens&quot;
## 9913 &quot;Bos taurus&quot;</code></pre>
<p><code>normalizeTaxa</code> does its best to figure out the order of
taxonomic levels automatically but can sometimes be left with ambiguous
cases. This will result in an error like:</p>
<pre><code>Error in topoSort(c(nonClade, list(lineageOrder)), errorIfAmbiguous = TRUE) : 
  Ambiguous ordering found in topoSort (suborder vs infraorder)</code></pre>
<p>That’s saying that the algorithm is unclear from the data whether
suborder or infraorder is the more specific taxonomic level. To clarify,
give the <code>lineageOrder</code> parameter a vector going from most to
least specific like:</p>
<pre><code>normalizeTaxa(raw,lineageOrder=c(&#39;infraorder&#39;,&#39;suborder&#39;))</code></pre>
<p>For especially troublesome sets, you may have to repeat this step
several times getting a new error each time to find all the ambiguities.
This would result in building up a vector specifying the ordering of
several ambiguous levels like:</p>
<pre><code>normalizeTaxa(raw,lineageOrder=c(&#39;infraorder&#39;,&#39;suborder&#39;,&#39;superorder&#39;,&#39;infraclass&#39;,&#39;subclass&#39;,&#39;class&#39;))
</code></pre>
</div>
<div id="finding-accessions-for-a-given-taxonomic-id" class="section level3">
<h3>Finding accessions for a given taxonomic ID</h3>
<p>To find all the accessions for a given taxonomic ID, you can use the
<code>getAccessions</code> function. This is a bit of an unusual use
case so to preserve space, an index is not created by default in
<code>read.accession2taxid</code>. If you are going to use this
function, you will want to rebuild the SQLite database with the
<code>indexTaxa</code> argument set to true with something like:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a><span class="fu">read.accession2taxid</span>(<span class="fu">list.files</span>(<span class="st">&#39;.&#39;</span>,<span class="st">&#39;accession2taxid.gz$&#39;</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>,<span class="at">indexTaxa=</span><span class="cn">TRUE</span>,<span class="at">overwrite=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## Reading nucl_gb.accession2taxid.gz.</code></pre>
<pre><code>## Reading nucl_wgs.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<p>Then you can get the accessions for taxa 3702 with a command like
(note that the limit argument is used here in order to preserve
space):</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" tabindex="-1"></a><span class="fu">getAccessions</span>(<span class="dv">3702</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>,<span class="at">limit=</span><span class="dv">10</span>)</span></code></pre></div>
<pre><code>##    taxa accession
## 1  3702  X58148.1
## 2  3702  X66414.1
## 3  3702  X60045.1
## 4  3702  X07376.1
## 5  3702  X54927.1
## 6  3702  X54926.1
## 7  3702  X54928.1
## 8  3702  X54930.1
## 9  3702  X54929.1
## 10 3702  X52320.1</code></pre>
</div>
<div id="convert-taxonomy-to-newick-tree" class="section level3">
<h3>Convert taxonomy to Newick tree</h3>
<p>This is probably only useful in a few specific cases but a
convenience function <code>makeNewick</code> to convert taxonomy into a
Newick tree is included. The function takes a matrix with columns
corresponding to taxonomic categories and rows corresponding to
taxonomic assignments, e.g. the output from <code>condenseTaxa</code> or
<code>getTaxonomy</code> or <code>normalizeTaxa</code> and reduces it to
a Newick formatted tree. For example:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a>taxa</span></code></pre></div>
<pre><code>##      [,1]        [,2]       [,3]       [,4]       [,5]        [,6]   
## [1,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot; 
## [2,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Pan&quot;  
## [3,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; NA         &quot;Cervidae&quot;  &quot;Alces&quot;</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="fu">makeNewick</span>(taxa)</span></code></pre></div>
<pre><code>## [1] &quot;((((((Homo,Pan)Hominidae)Primates,((Alces)Cervidae)_)Mammalia)Chordata)Eukaryota);&quot;</code></pre>
<p>If quotes are needed, then specify the <code>quote</code>
argument:</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" tabindex="-1"></a><span class="fu">makeNewick</span>(taxa,<span class="at">quote=</span><span class="st">&quot;&#39;&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;((((((&#39;Homo&#39;,&#39;Pan&#39;)&#39;Hominidae&#39;)&#39;Primates&#39;,((&#39;Alces&#39;)&#39;Cervidae&#39;)_)&#39;Mammalia&#39;)&#39;Chordata&#39;)&#39;Eukaryota&#39;);&quot;</code></pre>
<p>By default, <code>makeNewick</code> includes trailing nodes that are
all NA in the tree e.g.:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a>taxa[<span class="dv">3</span>,<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb80-2"><a href="#cb80-2" tabindex="-1"></a><span class="fu">print</span>(taxa)</span></code></pre></div>
<pre><code>##      [,1]        [,2]       [,3]       [,4]       [,5]        [,6]  
## [1,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Homo&quot;
## [2,] &quot;Eukaryota&quot; &quot;Chordata&quot; &quot;Mammalia&quot; &quot;Primates&quot; &quot;Hominidae&quot; &quot;Pan&quot; 
## [3,] &quot;Eukaryota&quot; &quot;Chordata&quot; NA         NA         NA          NA</code></pre>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" tabindex="-1"></a><span class="fu">makeNewick</span>(taxa)</span></code></pre></div>
<pre><code>## [1] &quot;((((((Homo,Pan)Hominidae)Primates)Mammalia,(((_)_)_)_)Chordata)Eukaryota);&quot;</code></pre>
<p>If these nodes are not desired then set
<code>excludeTerminalNAs</code> to <code>FALSE</code>:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" tabindex="-1"></a><span class="fu">makeNewick</span>(taxa,<span class="at">excludeTerminalNAs=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## [1] &quot;((((((Homo,Pan)Hominidae)Primates)Mammalia)Chordata)Eukaryota);&quot;</code></pre>
<p>Note that taxa may be the most specific taxon for a given taxa in the
taxonomy matrix but will not be a leaf in the resulting tree if it
appears in other taxonomy e.g. Chordata in this example.</p>
</div>
</div>
<div id="note-ncbi-name-changes-in-early-2023" class="section level2">
<h2>Note: NCBI name changes in early 2023</h2>
<p>Please note that the <a href="https://ncbiinsights.ncbi.nlm.nih.gov/2022/11/14/prokaryotic-phylum-name-changes/">NCBI
change their naming of several major prokaryote phylums</a> e.g. <a href="https://ftp.ncbi.nih.gov/pub/taxonomy/Major_taxonomic_updates_2023.txt">Firmicutes
became Bacillota</a> in early 2023. Please watch out for any problems
that could arise. For example: * names of assigned taxonomy may shift
after updating a database to a post-change version * comparisons of old
analyses performed pre-change to new analyses performed post-change will
need to be done with care</p>
<p>If I understand things correctly, then the actual taxonomy ID will
not change so it might be wise to retain the taxonomy ID for all
analyses. Then on final analysis, the taxonomic names can be assigned
based on whatever naming scheme is in use at that time.</p>
</div>
<div id="changelog" class="section level2">
<h2>Changelog</h2>
<div id="v0.10.7" class="section level3">
<h3>v0.10.7</h3>
<ul>
<li>Store <code>nodes</code> table’s <code>parent</code> column as INT
instead of REAL</li>
</ul>
</div>
<div id="v0.10.6" class="section level3">
<h3>v0.10.6</h3>
<ul>
<li>Fix test for error message that was variable across platforms</li>
</ul>
</div>
<div id="v0.10.5" class="section level3">
<h3>v0.10.5</h3>
<ul>
<li>Catch 404 errors and report as errors</li>
<li>Add resume argument to download functions</li>
<li>Don’t retain temp files for downloads if less than 10kb</li>
<li>README touchups</li>
</ul>
</div>
<div id="v0.10.4" class="section level3">
<h3>v0.10.4</h3>
<ul>
<li>Minor improvement to output md5 and modification date for downloads
to aid in debugging network issues</li>
</ul>
</div>
<div id="v0.10.3" class="section level3">
<h3>v0.10.3</h3>
<ul>
<li>Minor fix to prevent <code>accessionToTaxa</code> from hanging when
given numeric inputs</li>
</ul>
</div>
<div id="v0.10.2" class="section level3">
<h3>v0.10.2</h3>
<ul>
<li>Behind the scenes switch to <code>multi_download</code> function
from <code>curl</code> package to allow download resumption on
interrupted downloads. This adds a dependency that <code>curl</code>
package be &gt;=5.0.0.</li>
<li>Add <code>protocol</code> option to choose between FTP and HTTP
protocols for downloading. The two protocols should perform similarly
and the relative speeds of NCBI’s ftp and http servers seem to vary so
probably not a whole lot of reason to choose one over the other unless a
firewall is blocking FTP ports.</li>
</ul>
</div>
<div id="v0.10.1" class="section level3">
<h3>v0.10.1</h3>
<ul>
<li>Add <code>getDescendants</code> function to get all descendants for
a given taxon</li>
</ul>
</div>
<div id="v0.9.4" class="section level3">
<h3>v0.9.4</h3>
<ul>
<li>Add <code>getCommon</code> function to get all names in the database
for a given taxa ID</li>
</ul>
</div>
<div id="v0.9.3" class="section level3">
<h3>v0.9.3</h3>
<ul>
<li>Fix bug in testing script</li>
</ul>
</div>
<div id="v0.9.2" class="section level3">
<h3>v0.9.2</h3>
<ul>
<li>Allow factors as input to <code>accessionToTaxa</code></li>
<li>Document sqlite pragmas for <code>read.accession2taxid</code></li>
<li>Inherit … argument documentation for
<code>prepareDatabase</code></li>
<li>Catch input/output error while processing large files</li>
<li>Update various user-facing links from ftp to https for easier
access</li>
</ul>
</div>
<div id="v0.8.4" class="section level3">
<h3>v0.8.4</h3>
<ul>
<li>Add quote option to <code>makeNewick</code></li>
<li>Trim trailing NAs off the tree in <code>makeNewick</code> if
<code>excludeTerminalNAs</code> is TRUE</li>
<li>Add terminal semicolon to end of <code>makeNewick</code> tree unless
<code>terminator</code> is NULL</li>
</ul>
</div>
<div id="v0.8.3" class="section level3">
<h3>v0.8.3</h3>
<ul>
<li>Add “no rank” to <code>normalizeTaxa</code>’s default exclusion</li>
<li>Expand README</li>
</ul>
</div>
<div id="v0.8.2" class="section level3">
<h3>v0.8.2</h3>
<ul>
<li>Add <code>normalizeTaxa</code> function</li>
</ul>
</div>
<div id="v0.8.1" class="section level3">
<h3>v0.8.1</h3>
<ul>
<li>Fix minor typos</li>
</ul>
</div>
<div id="v0.8.0" class="section level3">
<h3>v0.8.0</h3>
<ul>
<li>Switch to <code>curl::curl_download</code> to avoid Windows
issues</li>
</ul>
</div>
<div id="v0.7.1" class="section level3">
<h3>v0.7.1</h3>
<ul>
<li>Add md5 check for downloads</li>
</ul>
</div>
<div id="v0.7.0" class="section level3">
<h3>v0.7.0</h3>
<ul>
<li>Add <code>getRawTaxonomy</code> function</li>
<li>Add option to not download accessions</li>
</ul>
</div>
<div id="v0.6.0" class="section level3">
<h3>v0.6.0</h3>
<ul>
<li>Fix named vector bug in <code>accessionToTaxa</code></li>
<li>Add <code>makeNewick</code> function</li>
<li>Deal with default 60 second timeout for downloads in R</li>
</ul>
</div>
<div id="v0.5.3" class="section level3">
<h3>v0.5.3</h3>
<ul>
<li>Remove <code>nucl_est</code> and <code>nucl_gss</code> from defaults
since NCBI folded them into <code>nucl_gb</code> and removed</li>
<li>Squash R:devel bug</li>
</ul>
</div>
<div id="v0.5.0" class="section level3">
<h3>v0.5.0</h3>
<ul>
<li>Transitioned from data.table to SQLite</li>
<li>Addeded convenience <code>prepareDatabase()</code> function</li>
<li>Squashed Windows testing errors</li>
</ul>
</div>
</div>
<div id="manual-preparation-of-database-usually-not-necessary" class="section level2">
<h2>Manual preparation of database (usually not necessary)</h2>
<p><strong>Note:</strong> Since version 0.5.0, it is usually not
necessary to run the following manually, the function
<code>prepareDatabase()</code> should do most of this automatically for
you (see <a href="#preparation">above</a>).</p>
<p>In order to avoid constant internet access and slow APIs, the first
step in using the package is to downloads all necessary files from NCBI.
This uses a bit of disk space but makes future access reliable and
fast.</p>
<p><strong>Note:</strong> It is not necessary to manually check for the
presence of these files since the functions automatically check to see
if their output is present and if so skip downloading/processing. Delete
the local files if you would like to redownload or reprocess them.</p>
<div id="download-names-and-nodes" class="section level3">
<h3>Download names and nodes</h3>
<p>First, download the necessary names and nodes files from <a href="https://ftp.ncbi.nih.gov/pub/taxonomy/">NCBI</a>:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a><span class="fu">getNamesAndNodes</span>()</span></code></pre></div>
<pre><code>## [1] &quot;./names.dmp&quot; &quot;./nodes.dmp&quot;</code></pre>
</div>
<div id="download-accession-to-taxa-files" class="section level3">
<h3>Download accession to taxa files</h3>
<p>Then download accession to taxa id conversion files from <a href="https://ftp.ncbi.nih.gov/pub/taxonomy/accession2taxid/">NCBI</a>.
<strong>Note:</strong> this is a pretty <em>big</em> download (several
gigabytes):</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a><span class="co">#this is a big download</span></span>
<span id="cb88-2"><a href="#cb88-2" tabindex="-1"></a><span class="fu">getAccession2taxid</span>()</span></code></pre></div>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## [1] &quot;./nucl_gb.accession2taxid.gz&quot;  &quot;./nucl_wgs.accession2taxid.gz&quot;</code></pre>
<p>If you would also like to identify protein accession numbers, also
download the prot file from NCBI (again this is a <em>big</em>
download):</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" tabindex="-1"></a><span class="co">#this is a big download</span></span>
<span id="cb91-2"><a href="#cb91-2" tabindex="-1"></a><span class="fu">getAccession2taxid</span>(<span class="at">types=</span><span class="st">&#39;prot&#39;</span>)</span></code></pre></div>
<pre><code>## This can be a big (several gigabytes) download. Please be patient and use a fast connection.</code></pre>
<pre><code>## [1] &quot;./prot.accession2taxid.gz&quot;</code></pre>
</div>
<div id="convert-names-nodes-and-accessions-to-database" class="section level3">
<h3>Convert names, nodes and accessions to database</h3>
<p>Then process the downloaded names and nodes files into a more easily
accessed form:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" tabindex="-1"></a><span class="fu">read.names.sql</span>(<span class="st">&#39;names.dmp&#39;</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span>
<span id="cb94-2"><a href="#cb94-2" tabindex="-1"></a><span class="fu">read.nodes.sql</span>(<span class="st">&#39;nodes.dmp&#39;</span>,<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<p>Next process the downloaded accession files into the same database
(this one could take a while):</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" tabindex="-1"></a><span class="fu">read.accession2taxid</span>(<span class="fu">list.files</span>(<span class="st">&#39;.&#39;</span>,<span class="st">&#39;accession2taxid.gz$&#39;</span>),<span class="st">&#39;accessionTaxa.sql&#39;</span>)</span></code></pre></div>
<pre><code>## Reading nucl_gb.accession2taxid.gz.</code></pre>
<pre><code>## Reading nucl_wgs.accession2taxid.gz.</code></pre>
<pre><code>## Reading prot.accession2taxid.gz.</code></pre>
<pre><code>## Reading in values. This may take a while.</code></pre>
<pre><code>## Adding index. This may also take a while.</code></pre>
<p>Now everything should be ready for processing. All files are cached
locally and so the preparation is only required once (or whenever you
would like to update the data). It is not necessary to manually check
for the presence of these files since the functions automatically check
to see if their output is present and if so skip downloading/processing.
Delete the local files if you would like to redownload or reprocess
them.</p>
</div>
</div>
<div id="switch-from-data.table-to-sqlite" class="section level2">
<h2>Switch from data.table to SQLite</h2>
<p>Version 0.5.0 marked a change for name and node lookups from using
data.table to using SQLite. This was necessary to increase performance
(10-100x speedup for <code>getTaxonomy</code>) and create a simpler
interface (a single SQLite database contains all necessary data).
Unfortunately, this switch requires a couple breaking changes: *
<code>getTaxonomy</code> changes from
<code>getTaxonomy(ids,namesDT,nodesDT)</code> to
<code>getTaxonomy(ids,sqlFile)</code> * <code>getId</code> changes from
<code>getId(taxa,namesDT)</code> to <code>getId(taxa,sqlFile)</code> *
<code>read.names</code> is deprecated, instead use
<code>read.names.sql</code>. For example, instead of calling
<code>names&lt;-read.names(&#39;names.dmp&#39;)</code> in every session, simply
call <code>read.names.sql(&#39;names.dmp&#39;,&#39;accessionTaxa.sql&#39;)</code> once
(or use the convenient <code>prepareDatabase</code> as
<a href="#preparation">above</a>)). * <code>read.nodes</code> is
deprecated, instead use <code>read.names.sql</code>. For example.
instead of calling <code>nodes&lt;-read.names(&#39;nodes.dmp&#39;)</code> in
every session, simply call
<code>read.nodes.sql(&#39;nodes.dmp&#39;,&#39;accessionTaxa.sql&#39;)</code> once (or
use the convenient <code>prepareDatabase</code> as
<a href="#preparation">above</a>).</p>
<p>I’ve tried to ease any problems with this by overloading
<code>getTaxonomy</code> and <code>getId</code> to still function (with
a warning) if passed a data.table names and nodes argument and providing
a simpler <code>prepareDatabase</code> function for completing all setup
steps (hopefully avoiding direct calls to <code>read.names</code> and
<code>read.nodes</code> for most users).</p>
<p>I plan to eventually remove data.table functionality to avoid a split
codebase so please switch to the new SQLite format in all new code.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
